papis/commands/tag.py

```diff
@@ -0,0 +1,194 @@
+
+"""
+from typing import List, Optional, Tuple
+
+import click
+
+import papis.cli
+import papis.commands
+import papis.importer
+import papis.commands.update
+import papis.commands.doctor
+import papis.config
+import papis.document
+import papis.strings
+import papis.tui.utils
+
+logger = papis.logging.get_logger(__name__)
+
+
+@click.command("tag")
+@click.help_option("--help", "-h")
+@papis.cli.git_option()
+@papis.cli.bool_flag(
+    "-d",
+    "--drop",
+    help="Drop all tags.",
+    default=False,
+)
+@click.option(
+    "-p",
+    "--add",
+    "to_add",
+    help="Add a tag.",
+    multiple=True,
+    type=str,
+)
+@click.option(
+    "-r",
+    "--remove",
+    "to_remove",
+    help="Remove a tag.",
+    multiple=True,
+    type=str,
+)
+@click.option(
+    "-n",
+    "--rename",
+    "to_rename",
+    help="Rename a tag (<OLD-TAG NEW-TAG>).",
+    multiple=True,
+    type=(str, str),
+)
+@papis.cli.bool_flag("-b", "--batch", help="Batch mode, do not prompt or otherwise")
+@papis.cli.doc_folder_option()
+@papis.cli.all_option()
+@papis.cli.sort_option()
+@papis.cli.query_argument()
+def cli(
+    git: bool,
+    drop: bool,
+    to_add: List[str],
+    to_remove: List[str],
+    to_rename: List[Tuple[str, str]],
+    sort_field: Optional[str],
+    sort_reverse: bool,
+    batch: bool,
+    query: str,
+    doc_folder: Tuple[str, ...],
+    _all: bool,
+) -> None:
+    """
+    Add and delete document tags.
+    """
+
+    if not papis.config.getboolean("tag-ignore-warning"):
+        logger.warning(
+            "This command assumes that your tags are a list (rather than a string). "
+            + "If this is not correct, it will not split them correctly and may give"
+            + "unexpected results. Use `papis doctor -t normalize-tags --fix` to "
+            + "clean up. To disable this warning, set `tag-ignore-warning=true` in "
+            + "your configuration."
+        )
+        if not papis.tui.utils.confirm("Are you sure you want to continue?"):
+            return
+
+    # if do_tag:
+    documents = papis.cli.handle_doc_folder_query_all_sort(
+        query, doc_folder, sort_field, sort_reverse, _all
+    )
+
+    if not documents:
+        logger.warning(papis.strings.no_documents_retrieved_message)
+        return
+
+    run_append = papis.commands.update.run_append
+    run_remove = papis.commands.update.run_remove
+    run_drop = papis.commands.update.run_drop
+    run_rename = papis.commands.update.run_rename
+    run_update = papis.commands.update.run
+
+    key_types = papis.commands.doctor.get_key_type_check_keys()
+    key_types["tags"] = type([])
+
+    success = True
+    processed_documents = []
+    for document in documents:
+        ctx = papis.importer.Context()
+
+        ctx.data.update(document)
+        if drop and success:
+            run_drop(ctx.data, ["tags"])
+
+        if to_add and success:
+            # NOTE: without the str(), it gets typed as Literal, which is a problem
```

Comment:
I just ran this locally (i.e. remove all the `str`) with `python -m mypy papis` and it doesn't seem to give any errors. 

What configuration are you using? Can you remove the `str` so we can see if this shows up on the CI as well?

Reply:
Oh, my bad! I keep forgetting that the pyright lsp that I have running locally thinks certain things are problematic which actually pass our CI. This is one of them!

Reply:
Hm, yeah, the CI uses `mypy` not `pyright` for the type checking, so they might give different results in some cases. Although this is a weird one.. it's being very restrictive.